project/src/main/ui/chat/career-cutscene-library.gd

```diff
@@ -0,0 +1,309 @@
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 	cutscenes to choose between.
+##
+## Returns:
+## 	A set of chat keys for cutscenes the player has already seen. This includes the leaf nodes for cutscenes which
+## 	have been seen, and branch nodes where all child cutscenes have been seen.
+func _exhausted_chat_keys(chat_key_roots: Array) -> Dictionary:
+	var excluded_chat_keys := {}
+	
+	# branches are only exhausted when their leaves are exhausted, so leaves must be traversed first (postorder)
+	var postorder_chat_keys := _postorder_chat_keys(chat_key_roots)
+	var chat_key_queue := postorder_chat_keys.duplicate()
+	while chat_key_queue:
+		var chat_key: String = chat_key_queue.pop_front()
+		var chat_key_pair: Dictionary = _chat_key_pairs_by_preroll.get(chat_key, {})
+		var preroll: String = chat_key_pair.get("preroll", "")
+		var postroll: String = chat_key_pair.get("postroll", "")
+		if preroll and not PlayerData.chat_history.is_chat_finished(preroll):
+			# don't exclude; preroll cutscene hasn't been played
+			continue
+		if postroll and not PlayerData.chat_history.is_chat_finished(postroll):
+			# don't exclude; postroll cutscene hasn't been played
+			continue
+		elif _preroll_tree.has(chat_key):
+			# calculate whether all child cutscenes have been played
+			var all_children_exhausted := true
+			for child in _preroll_tree.get(chat_key):
+				var child_key := _child_key(chat_key_roots, chat_key, child)
+				if not excluded_chat_keys.has(child_key):
+					all_children_exhausted = false
+			if not all_children_exhausted:
+				# don't exclude; child cutscenes haven't been played
+				continue
+		
+		# exclude; neither this node nor its children contains unplayed cutscenes
+		excluded_chat_keys[chat_key] = true
+	
+	return excluded_chat_keys
+
+
+## Returns a list of all chat keys in _preroll_tree in preroll order (leaves first.)
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 	cutscenes to choose between.
+##
+## Returns:
+## 	An array of String chat keys in postorder (leaves first)
+func _postorder_chat_keys(chat_key_roots: Array) -> Array:
+	var postorder_chat_keys := []
+	var chat_key_queue := chat_key_roots.duplicate()
+	while chat_key_queue:
+		var chat_key: String = chat_key_queue.pop_front()
+		
+		var children: Array = _preroll_tree.get(chat_key, [])
+		for child in children:
+			var child_key := _child_key(chat_key_roots, chat_key, child)
+			chat_key_queue.push_front(child_key)
+		
+		postorder_chat_keys.push_front(chat_key)
+	return postorder_chat_keys
+
+
+## Filters the list of potential chat key pairs, excluding the specified chat keys.
+##
+## The returned list includes the earliest numeric keys in each branch (0, 1, 2 are always played in order) as well as
+## ALL alphabetic keys in each branch (a, b, c are ordered arbitrarily)
```

Comment:
This is a very useful comment, and belongs higher up -- probably at the top of the script.

Reply:
Added elaborate top-level comment, `These cutscenes are ordered such that numeric siblings are always played in ascending order (00, 10, 11, 20..) but...`